src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/TbsCertificate.cs

```diff
@@ -0,0 +1,243 @@
+                List<byte[][]> encodedExtensions = new List<byte[][]>(Extensions.Count);
+
+                // extensions[3]  Extensions OPTIONAL
+                //
+                // Since this doesn't say IMPLICIT, it will look like
+                //
+                // A3 [length]
+                //   30 [length]
+                //     First Extension
+                //     Second Extension
+                //     ...
+
+                // An interesting quirk of skipping null values here is that
+                // Extensions.Count == 0 => no extensions
+                // Extensions.ContainsOnly(null) => empty extensions list
+
+                HashSet<string> usedOids = new HashSet<string>(Extensions.Count);
+
+                foreach (X509Extension extension in Extensions)
+                {
+                    if (extension == null)
+                        continue;
+
+                    if (!usedOids.Add(extension.Oid.Value))
+                    {
+                        throw new InvalidOperationException(
+                            SR.Format(SR.Cryptography_CertReq_DuplicateExtension, extension.Oid.Value));
+                    }
+
+                    encodedExtensions.Add(extension.SegmentedEncodedX509Extension());
+                }
+
+                byte[][] extensionField = DerEncoder.ConstructSegmentedSequence(
+                    DerEncoder.ConstructSegmentedSequence(encodedExtensions));
+                extensionField[0][0] = DerSequenceReader.ContextSpecificConstructedTag3;
+
+                encodedFields.Add(extensionField);
+            }
+
+            return DerEncoder.ConstructSequence(encodedFields);
+        }
+
+        private byte[][] EncodeValidityField(DateTimeOffset validityField, string propertyName)
+        {
+            /* https://tools.ietf.org/html/rfc3280#section-4.1.2.5
+             4.1.2.5  Validity
+
+                The certificate validity period is the time interval during which the
+                CA warrants that it will maintain information about the status of the
+                certificate.  The field is represented as a SEQUENCE of two dates:
+                the date on which the certificate validity period begins (notBefore)
+                and the date on which the certificate validity period ends
+                (notAfter).  Both notBefore and notAfter may be encoded as UTCTime or
+                GeneralizedTime.
+
+                CAs conforming to this profile MUST always encode certificate
+                validity dates through the year 2049 as UTCTime; certificate validity
+                dates in 2050 or later MUST be encoded as GeneralizedTime.
+
+                The validity period for a certificate is the period of time from
+                notBefore through notAfter, inclusive.
+            */
+
+            DateTime utcValue = validityField.UtcDateTime;
+
+            // On the one hand, GeneralizedTime easily goes back to 1000, and possibly to 0000;
```

Comment:
Comment seems a bit off as there would be a century issue, as only YY is recorded for dates < 2050, correct?